In [1]:
from sklearn import svm
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import pickle
from sklearn.metrics import classification_report, confusion_matrix

In [2]:
a = [1,2,3,4,5,6]
a[:3]

[1, 2, 3]

In [3]:
def loadCSV(filename, size):
    n = np.zeros(shape = (size, image_size*image_size))
    nn = np.zeros(shape = (size, 1))
    
    f = open(dest_folder+filename, 'r')
    i = 0
    for x in f:
        lst = x.split(',')
        nn[i] = np.asarray([lst[label]], dtype = np.float32)
        n[i] = np.asarray(lst, dtype=np.float32)[:image_size*image_size]
        i += 1
    for i in range(size):
        for j in range(image_size*image_size):
            n[i][j] /= 255
    return n, nn

In [4]:
dest_folder = './fashion_mnist/'
train_filename = 'train.csv'
test_filename = 'test.csv'
val_filename = 'val.csv'

train_size = 22500
test_size = 5000
val_size = 2501
image_size = 28
label = 784
num_classes = 10

In [5]:
X1, Y1 = loadCSV(train_filename, train_size)
X2, Y2 = loadCSV(test_filename, test_size)
X3, Y3 = loadCSV(val_filename, val_size)

In [7]:
clf = svm.SVC(kernel = 'rbf', gamma = 0.05)
clf.fit(X1, Y1.reshape(1,train_size)[0])

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma=0.05, kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [8]:
with open('q2_2_b.pkl', 'wb') as fid:
    pickle.dump(clf, fid)

In [9]:
Y2_pred = clf.predict(X2)
Y3_pred = clf.predict(X3)

In [32]:
print('===========TEST REPORT============')
print(confusion_matrix(Y2,Y2_pred))
print(classification_report(Y2,Y2_pred))

print('===========VALIDATION REPORT============')
print(confusion_matrix(Y3,Y3_pred))
print(classification_report(Y3,Y3_pred))

In [31]:
count_test = np.zeros(num_classes)
for i in range(val_size):
#     print(Y3[i][0])
    count_test[int(Y3[i][0])] += 1
# count_test

q2_2_c = np.zeros(num_classes)
for i in range(num_classes):
    q2_2_c[i] = (confusion_matrix(Y3, Y3_pred)[i][i])/count_test[i]

print('ACCURACY FOR VALIDATION DATA: ',q2_2_c) # [0.84462151 0.948 0.824 0.912 0.8 0.964 0.66 0.92 0.976 0.94] THUS, CLASS 6 WAS MISCLASSIFIED MOST

count_test = np.zeros(num_classes)
for i in range(test_size):
#     print(Y2[i][0])
    count_test[int(Y2[i][0])] += 1
# count_test

q2_2_c = np.zeros(num_classes)
for i in range(num_classes):
    q2_2_c[i] = (confusion_matrix(Y2, Y2_pred)[i][i])/count_test[i]

print('ACCURACY FOR TEST DATA: ',q2_2_c) # [0.866 0.964 0.822 0.914 0.798 0.946 0.63  0.942 0.978 0.948] THUS, CLASS 6 WAS MISCLASSIFIED MOST

ACCURACY FOR VALIDATION DATA:  [0.84462151 0.948      0.824      0.912      0.8        0.964
 0.66       0.92       0.976      0.94      ]
ACCURACY FOR TEST DATA:  [0.866 0.964 0.822 0.914 0.798 0.946 0.63  0.942 0.978 0.948]


==============================REPORT=================================

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma=0.05, kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)


===========TEST REPORT============
[[433   0   5  11   3   0  38   0  10   0]
 [  1 482   4   9   0   0   4   0   0   0]
 [  5   0 411   7  37   0  32   0   8   0]
 [ 12   0   3 457   9   0  14   0   5   0]
 [  3   1  41  13 399   0  38   0   5   0]
 [  0   0   0   0   0 473   0  16   5   6]
 [ 80   0  55   9  34   0 315   0   7   0]
 [  0   0   0   0   0  14   0 471   1  14]
 [  1   0   1   1   2   2   2   2 489   0]
 [  0   0   0   0   0  11   0  14   1 474]]
              precision    recall  f1-score   support

         0.0       0.81      0.87      0.84       500
         1.0       1.00      0.96      0.98       500
         2.0       0.79      0.82      0.81       500
         3.0       0.90      0.91      0.91       500
         4.0       0.82      0.80      0.81       500
         5.0       0.95      0.95      0.95       500
         6.0       0.71      0.63      0.67       500
         7.0       0.94      0.94      0.94       500
         8.0       0.92      0.98      0.95       500
         9.0       0.96      0.95      0.95       500

    accuracy                           0.88      5000
   macro avg       0.88      0.88      0.88      5000
weighted avg       0.88      0.88      0.88      5000

===========VALIDATION REPORT============
[[212   0   1   8   0   1  26   0   3   0]
 [  0 237   3   7   0   0   2   0   1   0]
 [  5   0 206   3  18   0  13   0   5   0]
 [  6   0   0 228   6   0   9   0   1   0]
 [  1   1  24   8 200   0  15   0   1   0]
 [  0   0   0   1   0 241   0   2   1   5]
 [ 34   0  28   3  19   0 165   0   1   0]
 [  0   0   0   0   0   8   0 230   1  11]
 [  0   0   1   1   1   0   1   2 244   0]
 [  0   0   0   0   0   6   0   8   1 235]]
              precision    recall  f1-score   support

         0.0       0.82      0.84      0.83       251
         1.0       1.00      0.95      0.97       250
         2.0       0.78      0.82      0.80       250
         3.0       0.88      0.91      0.90       250
         4.0       0.82      0.80      0.81       250
         5.0       0.94      0.96      0.95       250
         6.0       0.71      0.66      0.69       250
         7.0       0.95      0.92      0.93       250
         8.0       0.94      0.98      0.96       250
         9.0       0.94      0.94      0.94       250

    accuracy                           0.88      2501
   macro avg       0.88      0.88      0.88      2501
weighted avg       0.88      0.88      0.88      2501